# Практическая работа №4: Элементы корреляционного анализа. Проверка статистической гипотезы о равенстве коэффициента корреляции нулю

Выполнили студентки гр. 2384 Лавренова Юлия и Соц Екатерина. Вариант №32

## Цель работы
Освоение основных понятий, связанных с корреляционной зависимостью между случайными величинами, статистическими гипотезами и проверкой их «справедливости».

## Основные теоретические положения
Есть система двух случайных величин $\{X,Y\}$.

Эти случайные величины могут быть **независимыми**: $f(x,y)=f_1(x)\cdot f_2(y)$

В противном случае между ними может быть
- **функциональная зависимость**: $y=g(x)$
- **статистическая зависимость:**
$$\phi(x/y)=f(x,y)/f_2(y),$$ $$\phi(y/x)=f(x,y)/f_1(x)$$

**Корреляционной** называют статистическую зависимость случайных величин, при которой изменение значения одной из случайных величин приводит к изменению математического ожидания другой случайной величины.
$$M(X/y) = q_1(y),$$ $$M(Y/x) = q_2(x),$$

**Корреляционный момент:** $$\mu_{xy}=M\{[x-M(X)]\cdot [y-M(Y)]\}$$

**Коэффициент корреляции:** $$r_{xy}=\frac{\mu_{xy}}{\sigma_x \sigma_y}, |r_{xy}| \leq 1$$

Случайные величины $X$ и $Y$ называют коррелированными, если из корреляционный момент или их коэффициент корреляции отличен от нуля.

$X$ и $Y$ коррелированы $=>$ $X$ и $Y$ зависимы, $X$ и $Y$ независимы $=>$ $X$ и $Y$ некоррелированы.

Значение $\bar r_{xy}$ - статистической оценки $r_{xy}$ - коэффициента корреляции можно вычислить по формуле: $$\bar r_{xy} = \frac{\sum_{i=1}^{K_y}\sum_{j=1}^{K_x}n_{ij}y_i x_j - N \bar x_в \bar y_в}{NS_x S_y}$$
При N>50 в случае нормального распределения системы случайных величин $\{X,Y\}$ для оценки значения $r_{xy}$ можно использовать соотношение: $$\bar r_{xy} - 3\frac{1-\bar r_{xy}^2}{\sqrt N} \leq r_{xy} \leq \bar r_{xy} + 3\frac{1+\bar r_{xy}^2}{\sqrt N}$$
При вычислении выборочного коэффициента корреляции удобно использовать условные варианты: $$\bar r_{xy} = \frac{\sum_{i=1}^{K_y}\sum_{j=1}^{K_x}n_{ij}u_i v_j - N \bar u_в \bar v_в}{N\bar\sigma_u \bar\sigma_v}, u_i = \frac{x_i - C_x}{h_x}, v_j = \frac{y_j - C_y}{h_y}$$

**Доверительный интервал для выборочного коэффициента корреляции** генеральной совокупности с доверительной вероятностью $\gamma$ определяется по следующей схеме:
1. Вычисляется выборочное значение $\bar z$: $$\bar z = 0.5 \ln \frac{1+\bar r_{xy}}{1-\bar r_{xy}}=1.1513 \lg \frac{1+\bar r_{xy}}{1-\bar r_{xy}}$$
2. Вычисляется значение $\bar \sigma_z$: $$\bar \sigma_z =\frac{1}{\sqrt{N-3}}$$
3. Доверительный интервал для генерального значения представляется в виде: $$(\bar z - \lambda (\gamma)\bar \sigma_z, \bar z + \lambda (\gamma)\bar \sigma_z),$$
где значение $\lambda(\gamma)$ должно удовлетворять условию: $Ф[\lambda(\gamma)] = \frac{\gamma}{2}$.
4. Для пересчета интервала из п.3 в доверительный интервал для коэффициента корреляции с тем же значением $\gamma$ необходимо воспользоваться обратным преобразованием Фишера: $$r=th(z)=\frac{e^z - e^{-z}}{e^z + e^{-z}}=\frac{e^{2z}-1}{e^{2z}+1}$$

**Статистическая гипотеза**

Проверяется гипотеза о равенстве нулю генерального коэффициента корреляции: $$H_0: r_{xy} = 0$$
против альтернативы
$$H_1:r_{xy} \neq 0$$

Для этого вычисляется выборочный коэффициент корреляции $r_{xy}$ и статистика Стьюдента $t = \frac{r_{xy}\sqrt{n-2}}{\sqrt{1-r_{xy}^2}}$, которая при справедливости $H_0$ имеет распределение Стьюдента с $n-2$ степенями свободы.
Гипотеза $H_0$ отвергается, если $|t|$ превышает критическое значение.

## Постановка задачи
Из заданной генеральной совокупности сформировать выборку по второму признаку. Провести статистическую обработку второй выборки в объёме практических работ №1 и №2, с целью определения точечных статистических оценок параметров распределения исследуемого признака (математического ожидания, дисперсии, среднеквадратичного отклонения, асимметрии, эксцесса и коэффициента вариации). Для системы двух случайных величин X (первый признак) и Y (второй признак) сформировать двумерную выборку и найти статистическую оценку коэффициента корреляции, построить доверительный интервал для коэффициента корреляции и осуществить проверку статистической гипотезы о равенстве коэффициента корреляции нулю. Полученные результаты содержательно проинтерпретировать.

## Выполнение работы

### **1.** Провести статистическую обработку второй выборки в объёме практических работ №1 и №2, с целью определения точечных статистических оценок параметров распределения исследуемого признака (математического ожидания, дисперсии, среднеквадратичного отклонения, асимметрии, эксцесса, моды, медианы и коэффициента вариации). Оформить результаты в виде таблицы, сделать выводы.

Первая и вторая практические работы уже содержат в себе статистическую обработку для обоих признаков. Соберем все в одну таблицу:

In [ ]:
import pandas as pd
import numpy as np
from IPython.display import display, Markdown

data = pd.read_csv("sample_wine_data.csv")

def group_array(data):
    # ранжированный ряд
    df_range = np.sort(data)

    # вариационный ряд
    df_unique = np.unique(df_range)
    df_var = {value: np.sum(df_range == value) for value in df_unique}

    # интервальный ряд
    N = len(data)
    k = round(1 + 3.322 * np.log10(N))
    bins = np.linspace(min(df_unique), max(df_unique), k)
    m_i, edges = np.histogram(data, bins=bins)

    x_i_mid = (edges[:-1] + edges[1:]) / 2
    m_i_rel = m_i / N
    m_i_cum = np.cumsum(m_i)
    m_i_rel_cum = np.cumsum(m_i_rel)

    table = pd.DataFrame({
        "i": range(1, len(m_i) + 1),
        "$[x_i, x_{i+1})$": [f"[{edges[i]:.2f}, {edges[i+1]:.2f})" for i in range(len(m_i))],
        "$\\tilde{x}_i$": np.round(x_i_mid, 3),
        "$m_i$": m_i,
        "$\\tilde{m}_i$": np.round(m_i_rel, 4),
        "$m_{i}^{нак}$": m_i_cum,
        "$\\tilde{m}_{i}^{нак}$": np.round(m_i_rel_cum, 4)
    })

    sum_row = pd.DataFrame({
        "i": ["Σ"],
        "$[x_i, x_{i+1})$": ["-"],
        "$\\tilde{x}_i$": ["-"],
        "$m_i$": [m_i.sum()],
        "$\\tilde{m}_i$": [m_i_rel.sum()],
        "$m_{i}^{нак}$": ["-"],
        "$\\tilde{m}_{i}^{нак}$": ["-"]
    })

    table = pd.concat([table, sum_row], ignore_index=True)

    return df_range, df_var, table

# Обработка pH
data_ph = np.array(data["pH"])
ph_range, ph_var, ph_table = group_array(data_ph)

print("\nИнтервальный ряд для pH:")
display(Markdown(ph_table.to_markdown(index=False)))

# Обработка sulphates
data_sulphates = np.array(data["sulphates"])
sulphates_range, sulphates_var, sulphates_table = group_array(data_sulphates)

print("\nИнтервальный ряд для Sulphates:")
display(Markdown(sulphates_table.to_markdown(index=False)))


# Извлекаем данные из таблиц (исключая строку с суммой)
ph_data = ph_table.iloc[:-1].copy()
ph_data['i'] = ph_data['i'].astype(int)
ph_data['$\\tilde{x}_i$'] = ph_data['$\\tilde{x}_i$'].astype(float)
ph_data['$m_i$'] = ph_data['$m_i$'].astype(int)

sulphates_data = sulphates_table.iloc[:-1].copy()
sulphates_data['i'] = sulphates_data['i'].astype(int)
sulphates_data['$\\tilde{x}_i$'] = sulphates_data['$\\tilde{x}_i$'].astype(float)
sulphates_data['$m_i$'] = sulphates_data['$m_i$'].astype(int)

# 1. ВЫЧИСЛЕНИЕ УСЛОВНЫХ ВАРИАНТ
def calculate_conditional_variants(data_table, feature_name):
    df = data_table.copy()

    # Определяем шаг интервала h
    x_mid = df['$\\tilde{x}_i$'].values
    h = round(x_mid[1] - x_mid[0], 3)

    # Выбираем условный нуль C (середина интервала с максимальной частотой)
    max_freq_idx = df['$m_i$'].idxmax()
    C = df.loc[max_freq_idx, '$\\tilde{x}_i$']

    # Вычисляем условные варианты u_i
    df['$u_i$'] = ((df['$\\tilde{x}_i$'] - C) / h).round(3)

    # Вычисляем необходимые столбцы
    df['$n_i u_i$'] = (df['$m_i$'] * df['$u_i$']).round(3)
    df['$n_i u_i^2$'] = (df['$m_i$'] * df['$u_i$']**2).round(3)
    df['$n_i u_i^3$'] = (df['$m_i$'] * df['$u_i$']**3).round(3)
    df['$n_i u_i^4$'] = (df['$m_i$'] * df['$u_i$']**4).round(3)
    df['$n_i (u_i+1)^4$'] = (df['$m_i$'] * (df['$u_i$'] + 1)**4).round(3)

    result_table = df[['i', '$\\tilde{x}_i$', '$m_i$', '$u_i$',
                       '$n_i u_i$', '$n_i u_i^2$', '$n_i u_i^3$',
                       '$n_i u_i^4$', '$n_i (u_i+1)^4$']].copy()

    sums = pd.DataFrame({
        'i': ['Σ'],
        '$\\tilde{x}_i$': ['-'],
        '$m_i$': [df['$m_i$'].sum()],
        '$u_i$': ['-'],
        '$n_i u_i$': [result_table['$n_i u_i$'].sum()],
        '$n_i u_i^2$': [result_table['$n_i u_i^2$'].sum()],
        '$n_i u_i^3$': [result_table['$n_i u_i^3$'].sum()],
        '$n_i u_i^4$': [result_table['$n_i u_i^4$'].sum()],
        '$n_i (u_i+1)^4$': [result_table['$n_i (u_i+1)^4$'].sum()]
    })

    final_table = pd.concat([result_table, sums], ignore_index=True)

    return final_table, h, C, df['$m_i$'].sum()

# Вычисляем для pH
ph_conditional_table, ph_h, ph_C, ph_n = calculate_conditional_variants(ph_data, "pH")
print("\nТаблица условных вариант для pH:")
display(Markdown(ph_conditional_table.to_markdown(index=False)))

# Вычисляем для sulphates
sulphates_conditional_table, sulphates_h, sulphates_C, sulphates_n = calculate_conditional_variants(sulphates_data, "SULPHATES")
print("\nТаблица условных вариант для Sulphates:")
display(Markdown(sulphates_conditional_table.to_markdown(index=False)))

# 2. ВЫЧИСЛЕНИЕ МОМЕНТОВ
def calculate_moments(conditional_table, h, n, feature_name):
    print(f"\n--- Вычисление моментов для {feature_name} ---")

    sum_ni_ui = conditional_table[conditional_table['i'] != 'Σ']['$n_i u_i$'].sum()
    sum_ni_ui2 = conditional_table[conditional_table['i'] != 'Σ']['$n_i u_i^2$'].sum()
    sum_ni_ui3 = conditional_table[conditional_table['i'] != 'Σ']['$n_i u_i^3$'].sum()
    sum_ni_ui4 = conditional_table[conditional_table['i'] != 'Σ']['$n_i u_i^4$'].sum()

    # Условные эмпирические моменты
    M1_star = sum_ni_ui / n
    M2_star = sum_ni_ui2 / n
    M3_star = sum_ni_ui3 / n
    M4_star = sum_ni_ui4 / n

    print(f"M*₁ = {M1_star:.4f}")
    print(f"M*₂ = {M2_star:.4f}")
    print(f"M*₃ = {M3_star:.4f}")
    print(f"M*₄ = {M4_star:.4f}")

    # Центральные эмпирические моменты
    m1 = 0
    m2 = h**2 * (M2_star - M1_star**2)
    m3 = h**3 * (M3_star - 3*M1_star*M2_star + 2*M1_star**3)
    m4 = h**4 * (M4_star - 4*M1_star*M3_star + 6*M1_star**2 * M2_star - 3*M1_star**4)

    print(f"m₁ = {m1}")
    print(f"m₂ = {m2:.6f}")
    print(f"m₃ = {m3:.6f}")
    print(f"m₄ = {m4:.6f}")

    return {
        'M1_star': M1_star, 'M2_star': M2_star, 'M3_star': M3_star, 'M4_star': M4_star,
        'm1': m1, 'm2': m2, 'm3': m3, 'm4': m4,
        'h': h, 'n': n
    }

ph_moments = calculate_moments(ph_conditional_table, ph_h, ph_n, "pH")
sulphates_moments = calculate_moments(sulphates_conditional_table, sulphates_h, sulphates_n, "Sulphates")

# 3. ВЫЧИСЛЕНИЕ ВЫБОРОЧНЫХ ХАРАКТЕРИСТИК
def calculate_sample_characteristics(moments, C, raw_data, feature_name):
    print(f"\n--- Выборочные характеристики для {feature_name} ---")

    # Через условные варианты
    x_bar_cond = C + moments['h'] * moments['M1_star']
    D_cond = moments['m2']
    sigma_cond = np.sqrt(D_cond)

    print(f"Среднее (условные варианты): {x_bar_cond:.4f}")
    print(f"Дисперсия (условные варианты): {D_cond:.6f}")
    print(f"СКО (условные варианты): {sigma_cond:.6f}")

    # По стандартной формуле
    x_bar_std = np.mean(raw_data)
    D_std = np.var(raw_data, ddof=0)
    sigma_std = np.std(raw_data, ddof=0)

    print(f"Среднее (стандартная формула): {x_bar_std:.4f}")
    print(f"Дисперсия (стандартная формула): {D_std:.6f}")
    print(f"СКО (стандартная формула): {sigma_std:.6f}")

    # Сравнение
    print(f"Разница средних: {abs(x_bar_cond - x_bar_std):.6f}")
    print(f"Разница дисперсий: {abs(D_cond - D_std):.6f}")

    return {
        'x_bar': x_bar_cond,
        'D': D_cond,
        'sigma': sigma_cond,
        'x_bar_std': x_bar_std,
        'D_std': D_std,
        'sigma_std': sigma_std
    }

ph_results = calculate_sample_characteristics(ph_moments, ph_C, data_ph, "pH")
sulphates_results = calculate_sample_characteristics(sulphates_moments, sulphates_C, data_sulphates, "Sulphates")

# 4. ИСПРАВЛЕННЫЕ ОЦЕНКИ
def calculate_corrected_estimates(results, raw_data, feature_name):
    print(f"\n--- Исправленные оценки для {feature_name} ---")

    n = len(raw_data)
    D_biased = results['D']
    sigma_biased = results['sigma']

    s2 = (n / (n - 1)) * D_biased
    s = np.sqrt(s2)

    print(f"Объем выборки: n = {n}")
    print(f"Смещенная дисперсия: {D_biased:.6f}")
    print(f"Исправленная дисперсия: {s2:.6f}")
    print(f"Разница: {s2 - D_biased:.6f}")
    print(f"Смещенное СКО: {sigma_biased:.6f}")
    print(f"Исправленное СКО: {s:.6f}")
    print(f"Разница: {s - sigma_biased:.6f}")

    return {'s2': s2, 's': s}

ph_corrected = calculate_corrected_estimates(ph_results, data_ph, "pH")
sulphates_corrected = calculate_corrected_estimates(sulphates_results, data_sulphates, "Sulphates")

# 5. КОЭФФИЦИЕНТЫ АСИММЕТРИИ И ЭКСЦЕССА
def calculate_skewness_kurtosis(moments, feature_name):
    print(f"\n--- Коэффициенты асимметрии и эксцесса для {feature_name} ---")

    m2 = moments['m2']
    m3 = moments['m3']
    m4 = moments['m4']

    As = m3 / (m2 ** (3/2))
    E = (m4 / (m2 ** 2)) - 3

    print(f"m2 = {m2:.6f}")
    print(f"m3 = {m3:.6f}")
    print(f"m4 = {m4:.6f}")
    print(f"Коэффициент асимметрии As = {As:.6f}")
    print(f"Коэффициент эксцесса E = {E:.6f}")

    # Интерпретация
    if abs(As) < 0.5:
        print(" Асимметрия умеренная")
    elif abs(As) < 1:
        print(" Асимметрия заметная")
    else:
        print(" Асимметрия сильная")

    if E > 0:
        print(" Распределение островершинное")
    elif E < 0:
        print(" Распределение плосковершинное")
    else:
        print(" Распределение близко к нормальному")

    return {'As': As, 'E': E}

ph_skew_kurt = calculate_skewness_kurtosis(ph_moments, "pH")
sulphates_skew_kurt = calculate_skewness_kurtosis(sulphates_moments, "Sulphates")

# 6. МОДА, МЕДИАНА И КОЭФФИЦИЕНТ ВАРИАЦИИ
def calculate_mode_median_variation(interval_table, results, raw_data, feature_name):
    print(f"\n--- Мода, медиана и коэффициент вариации для {feature_name} ---")

    df = interval_table.iloc[:-1].copy()
    n = len(raw_data)

    # Извлекаем границы интервалов
    intervals = df["$[x_i, x_{i+1})$"].str.extract(r"\[(.*), (.*)\)").astype(float)
    df["L"] = intervals[0]
    df["R"] = intervals[1]

    # Шаг интервала
    h = df["R"].iloc[0] - df["L"].iloc[0]

    # МОДА
    modal_index = df["$m_i$"].idxmax()
    Lm = df.loc[modal_index, "L"]
    fm = df.loc[modal_index, "$m_i$"]

    f_prev = df.loc[modal_index - 1, "$m_i$"] if modal_index > 0 else 0
    f_next = df.loc[modal_index + 1, "$m_i$"] if modal_index < len(df) - 1 else 0

    Mo = Lm + h * ((fm - f_prev) / ((fm - f_prev) + (fm - f_next)))

    # МЕДИАНА
    df["cum_freq"] = df["$m_i$"].cumsum()
    median_index = df[df["cum_freq"] >= n/2].index[0]

    Lmed = df.loc[median_index, "L"]
    F_prev = df.loc[median_index - 1, "cum_freq"] if median_index > 0 else 0
    f_med = df.loc[median_index, "$m_i$"]

    Me = Lmed + h * ((n/2 - F_prev) / f_med)

    # КОЭФФИЦИЕНТ ВАРИАЦИИ
    sigma = results["sigma"]
    x_bar = results["x_bar"]
    V = (sigma / x_bar) * 100

    print(f"Мода Mo* = {Mo:.4f}")
    print(f"Медиана Me* = {Me:.4f}")
    print(f"Коэффициент вариации V* = {V:.2f}%")

    if V < 10:
        print(" Слабая вариация, совокупность однородна")
    elif V < 20:
        print(" Умеренная вариация")
    elif V < 33:
        print(" Значительная вариация, совокупность однородна")
    else:
        print(" Сильная вариация, совокупность неоднородна")

    return {'mode': Mo, 'median': Me, 'variation': V}

ph_distribution = calculate_mode_median_variation(ph_table, ph_results, data_ph, "pH")
sulphates_distribution = calculate_mode_median_variation(sulphates_table, sulphates_results, data_sulphates, "Sulphates")

# ИТОГОВАЯ ТАБЛИЦА
print("ИТОГОВАЯ ТАБЛИЦА ТОЧЕЧНЫХ СТАТИСТИЧЕСКИХ ОЦЕНОК")

final_table = pd.DataFrame({
    "Статистическая оценка": [
        "Математическое ожидание (среднее выборочное)",
        "Дисперсия выборочная",
        "Среднеквадратическое отклонение",
        "Исправленная дисперсия",
        "Исправленное СКО",
        "Коэффициент асимметрии",
        "Коэффициент эксцесса",
        "Мода",
        "Медиана",
        "Коэффициент вариации (%)"
    ],
    "pH": [
        f"{ph_results['x_bar']:.4f}",
        f"{ph_results['D']:.4f}",
        f"{ph_results['sigma']:.4f}",
        f"{ph_corrected['s2']:.4f}",
        f"{ph_corrected['s']:.4f}",
        f"{ph_skew_kurt['As']:.4f}",
        f"{ph_skew_kurt['E']:.4f}",
        f"{ph_distribution['mode']:.4f}",
        f"{ph_distribution['median']:.4f}",
        f"{ph_distribution['variation']:.2f}"
    ],
    "Sulphates": [
        f"{sulphates_results['x_bar']:.4f}",
        f"{sulphates_results['D']:.4f}",
        f"{sulphates_results['sigma']:.4f}",
        f"{sulphates_corrected['s2']:.4f}",
        f"{sulphates_corrected['s']:.4f}",
        f"{sulphates_skew_kurt['As']:.4f}",
        f"{sulphates_skew_kurt['E']:.4f}",
        f"{sulphates_distribution['mode']:.4f}",
        f"{sulphates_distribution['median']:.4f}",
        f"{sulphates_distribution['variation']:.2f}"
    ]
})

display(Markdown(final_table.to_markdown(index=False)))


Интервальный ряд для pH:


| i   | $[x_i, x_{i+1})$   | $\tilde{x}_i$   |   $m_i$ |   $\tilde{m}_i$ | $m_{i}^{нак}$   | $\tilde{m}_{i}^{нак}$   |
|:----|:-------------------|:----------------|--------:|----------------:|:----------------|:------------------------|
| 1   | [2.86, 3.01)       | 2.934           |       5 |          0.0439 | 5               | 0.0439                  |
| 2   | [3.01, 3.16)       | 3.083           |      12 |          0.1053 | 17              | 0.1491                  |
| 3   | [3.16, 3.31)       | 3.231           |      38 |          0.3333 | 55              | 0.4825                  |
| 4   | [3.31, 3.45)       | 3.38            |      41 |          0.3596 | 96              | 0.8421                  |
| 5   | [3.45, 3.60)       | 3.529           |      13 |          0.114  | 109             | 0.9561                  |
| 6   | [3.60, 3.75)       | 3.677           |       3 |          0.0263 | 112             | 0.9825                  |
| 7   | [3.75, 3.90)       | 3.826           |       2 |          0.0175 | 114             | 1.0                     |
| Σ   | -                  | -               |     114 |          1      | -               | -                       |


Интервальный ряд для Sulphates:


| i   | $[x_i, x_{i+1})$   | $\tilde{x}_i$   |   $m_i$ |   $\tilde{m}_i$ | $m_{i}^{нак}$   | $\tilde{m}_{i}^{нак}$   |
|:----|:-------------------|:----------------|--------:|----------------:|:----------------|:------------------------|
| 1   | [0.44, 0.61)       | 0.524           |      53 |          0.4649 | 53              | 0.4649                  |
| 2   | [0.61, 0.78)       | 0.693           |      42 |          0.3684 | 95              | 0.8333                  |
| 3   | [0.78, 0.95)       | 0.861           |      13 |          0.114  | 108             | 0.9474                  |
| 4   | [0.95, 1.11)       | 1.03            |       1 |          0.0088 | 109             | 0.9561                  |
| 5   | [1.11, 1.28)       | 1.199           |       2 |          0.0175 | 111             | 0.9737                  |
| 6   | [1.28, 1.45)       | 1.367           |       2 |          0.0175 | 113             | 0.9912                  |
| 7   | [1.45, 1.62)       | 1.536           |       1 |          0.0088 | 114             | 1.0                     |
| Σ   | -                  | -               |     114 |          1      | -               | -                       |


Таблица условных вариант для pH:


| i   | $\tilde{x}_i$   |   $m_i$ | $u_i$   |   $n_i u_i$ |   $n_i u_i^2$ |   $n_i u_i^3$ |   $n_i u_i^4$ |   $n_i (u_i+1)^4$ |
|:----|:----------------|--------:|:--------|------------:|--------------:|--------------:|--------------:|------------------:|
| 1   | 2.934           |       5 | -2.993  |     -14.965 |        44.79  |      -134.057 |       401.233 |            78.886 |
| 2   | 3.083           |      12 | -1.993  |     -23.916 |        47.665 |       -94.996 |       189.326 |            11.668 |
| 3   | 3.231           |      38 | -1.0    |     -38     |        38     |       -38     |        38     |             0     |
| 4   | 3.38            |      41 | 0.0     |       0     |         0     |         0     |         0     |            41     |
| 5   | 3.529           |      13 | 1.0     |      13     |        13     |        13     |        13     |           208     |
| 6   | 3.677           |       3 | 1.993   |       5.979 |        11.916 |        23.749 |        47.332 |           240.74  |
| 7   | 3.826           |       2 | 2.993   |       5.986 |        17.916 |        53.623 |       160.493 |           508.425 |
| Σ   | -               |     114 | -       |     -51.916 |       173.287 |      -176.681 |       849.384 |          1088.72  |


Таблица условных вариант для Sulphates:


| i   | $\tilde{x}_i$   |   $m_i$ | $u_i$   |   $n_i u_i$ |   $n_i u_i^2$ |   $n_i u_i^3$ |   $n_i u_i^4$ |   $n_i (u_i+1)^4$ |
|:----|:----------------|--------:|:--------|------------:|--------------:|--------------:|--------------:|------------------:|
| 1   | 0.524           |      53 | 0.0     |       0     |         0     |         0     |         0     |            53     |
| 2   | 0.693           |      42 | 1.0     |      42     |        42     |        42     |        42     |           672     |
| 3   | 0.861           |      13 | 1.994   |      25.922 |        51.688 |       103.067 |       205.515 |          1044.6   |
| 4   | 1.03            |       1 | 2.994   |       2.994 |         8.964 |        26.838 |        80.354 |           254.467 |
| 5   | 1.199           |       2 | 3.994   |       7.988 |        31.904 |       127.425 |       508.935 |          1244.01  |
| 6   | 1.367           |       2 | 4.988   |       9.976 |        49.76  |       248.204 |      1238.04  |          2571.33  |
| 7   | 1.536           |       1 | 5.988   |       5.988 |        35.856 |       214.707 |      1285.66  |          2384.58  |
| Σ   | -               |     114 | -       |      94.868 |       220.172 |       762.241 |      3360.51  |          8223.98  |


--- Вычисление моментов для pH ---
M*₁ = -0.4554
M*₂ = 1.5201
M*₃ = -1.5498
M*₄ = 7.4507
m₁ = 0
m₂ = 0.029143
m₃ = 0.001118
m₄ = 0.003150

--- Вычисление моментов для Sulphates ---
M*₁ = 0.8322
M*₂ = 1.9313
M*₃ = 6.6863
M*₄ = 29.4782
m₁ = 0
m₂ = 0.035382
m₃ = 0.014564
m₄ = 0.011263

--- Выборочные характеристики для pH ---
Среднее (условные варианты): 3.3121
Дисперсия (условные варианты): 0.029143
СКО (условные варианты): 0.170712
Среднее (стандартная формула): 3.3077
Дисперсия (стандартная формула): 0.031381
СКО (стандартная формула): 0.177146
Разница средних: 0.004426
Разница дисперсий: 0.002238

--- Выборочные характеристики для Sulphates ---
Среднее (условные варианты): 0.6646
Дисперсия (условные варианты): 0.035382
СКО (условные варианты): 0.188101
Среднее (стандартная формула): 0.6601
Дисперсия (стандартная формула): 0.035057
СКО (стандартная формула): 0.187235
Разница средних: 0.004550
Разница дисперсий: 0.000325

--- Исправленные оценки для pH ---
Объем выборки: n = 114
Смещен

| Статистическая оценка                        |     pH |   Sulphates |
|:---------------------------------------------|-------:|------------:|
| Математическое ожидание (среднее выборочное) | 3.3121 |      0.6646 |
| Дисперсия выборочная                         | 0.0291 |      0.0354 |
| Среднеквадратическое отклонение              | 0.1707 |      0.1881 |
| Исправленная дисперсия                       | 0.0294 |      0.0357 |
| Исправленное СКО                             | 0.1715 |      0.1889 |
| Коэффициент асимметрии                       | 0.2247 |      2.1883 |
| Коэффициент эксцесса                         | 0.7084 |      5.9971 |
| Мода                                         | 3.3245 |      0.5808 |
| Медиана                                      | 3.3173 |      0.6262 |
| Коэффициент вариации (%)                     | 5.15   |     28.3    |

1. МАТЕМАТИЧЕСКОЕ ОЖИДАНИЕ:
   - pH: 3.3121 - типичный уровень кислотности образцов вина
   - Sulphates: 0.6646 - среднее содержание сульфатов

2. ДИСПЕРСИЯ И СКО:
   - pH: $D$ = 0.0291, $\sigma$ = 0.1707 - умеренный разброс значений
   - Sulphates: $D$ = 0.0354, $\sigma$ = 0.1881 - несколько больший разброс

3. ИСПРАВЛЕННЫЕ ОЦЕНКИ:
   - Незначительно превышают смещенные (на 0.0003-0.0008)
   - При n=114 различия минимальны, но для выводов используем исправленные

4. АСИММЕТРИЯ:
   - pH: As = 0.2247 - умеренная положительная асимметрия
   - Sulphates: As = 2.1883 - сильная положительная асимметрия

5. ЭКСЦЕСС:
   - pH: E = 0.7084 - слабая островершинность
   - Sulphates: E = 5.9971 - резко выраженная островершинность

6. МОДА И МЕДИАНА:
   - pH: Mo ≈ Me ≈ x̄ - подтверждает симметричность
   - Sulphates: Mo < Me < x̄ - подтверждает правостороннюю асимметрию

7. КОЭФФИЦИЕНТ ВАРИАЦИИ:
   - pH: V = 5.15% - слабая вариация, высокая однородность
   - Sulphates: V = 28.30% - умеренная вариация, статистическая однородность


Признак pH характеризуется высокой однородностью и близостью к нормальному распределению.
Признак sulphates демонстрирует значительную асимметрию и островершинность при сохранении
статистической однородности совокупности. Полученные оценки могут быть использованы для
дальнейшего корреляционного анализа и построения доверительных интервалов.

### **2.** Построить двумерный интервальный вариационный ряд, оформить в виде таблицы.

In [ ]:
X = data['pH'].values
Y = data['sulphates'].values
n = len(X)

# Число интервалов по формуле Стёрджеса
k = int(np.ceil(1 + 3.322 * np.log10(n)))

# Границы интервалов
bins_x = np.linspace(X.min(), X.max(), k + 1)
bins_y = np.linspace(Y.min(), Y.max(), k + 1)

# Двумерная таблица частот
H, xedges, yedges = np.histogram2d(X, Y, bins=[bins_x, bins_y])
H = H.astype(int)

# Метки строк и столбцов
row_labels = [f'[{xedges[i]:.3f}; {xedges[i+1]:.3f})' for i in range(k)]
col_labels = [f'[{yedges[j]:.3f}; {yedges[j+1]:.3f})' for j in range(k)]


# Собираем DataFrame с итогами
df_table = pd.DataFrame(H, index=row_labels, columns=col_labels)
df_table['$n_y$'] = df_table.sum(axis=1)
df_table.loc['$n_x$'] = df_table.sum(axis=0)

# Заменяем нули на "—"
df_display = df_table.copy().astype(object)
for col in df_display.columns[:-1]:
    for idx in df_display.index[:-1]:
        if df_display.loc[idx, col] == 0:
            df_display.loc[idx, col] = '—'

df_display.index.name = 'pH \\ sulphates'
display(Markdown(df_display.to_markdown()))

| pH \ sulphates   | [0.440; 0.588)   | [0.588; 0.735)   | [0.735; 0.883)   | [0.883; 1.030)   | [1.030; 1.177)   | [1.177; 1.325)   | [1.325; 1.473)   | [1.473; 1.620)   |   $n_y$ |
|:-----------------|:-----------------|:-----------------|:-----------------|:-----------------|:-----------------|:-----------------|:-----------------|:-----------------|--------:|
| [2.860; 2.990)   | 2                | —                | 1                | —                | —                | —                | 1                | 1                |       5 |
| [2.990; 3.120)   | 4                | 1                | —                | 1                | 1                | —                | 1                | —                |       8 |
| [3.120; 3.250)   | 8                | 11               | 6                | 1                | —                | —                | —                | —                |      26 |
| [3.250; 3.380)   | 18               | 11               | 7                | —                | —                | 2                | —                | —                |      38 |
| [3.380; 3.510)   | 10               | 14               | 2                | —                | —                | —                | —                | —                |      26 |
| [3.510; 3.640)   | 2                | 4                | —                | —                | —                | —                | —                | —                |       6 |
| [3.640; 3.770)   | 1                | —                | 2                | —                | —                | —                | —                | —                |       3 |
| [3.770; 3.900)   | —                | 2                | —                | —                | —                | —                | —                | —                |       2 |
| $n_x$            | 45               | 43               | 18               | 2                | 1                | 2                | 2                | 1                |     114 |

### **3.** По полученному двумерному интервальному вариационному ряду построить корреляционную таблицу, сделать выводы.

In [ ]:
X = data['pH'].values
Y = data['sulphates'].values
n = len(X)

# Определяем количество интервалов
k = int(np.ceil(1 + 3.322 * np.log10(n)))

# Строим интервалы
bins_x = np.linspace(X.min(), X.max(), k + 1)
bins_y = np.linspace(Y.min(), Y.max(), k + 1)

# Вычисляем середины интервалов
mid_x = np.array([(bins_x[i] + bins_x[i+1]) / 2 for i in range(k)])
mid_y = np.array([(bins_y[j] + bins_y[j+1]) / 2 for j in range(k)])

# Строим двумерную гистограмму
H, xedges, yedges = np.histogram2d(X, Y, bins=[bins_x, bins_y])
H = H.astype(int)

# Создаем подписи для строк и столбцов (используем середины интервалов)
row_labels = [f'{mid_x[i]:.3f}' for i in range(k)]
col_labels = [f'{mid_y[j]:.3f}' for j in range(k)]

# Суммы по строкам и столбцам
row_sums = H.sum(axis=1)   # n_i
col_sums = H.sum(axis=0)   # n_j

# Взвешенные суммы
weighted_sum_y = H @ mid_y  # для каждой строки: сумма n_ij * y_j
weighted_sum_x = H.T @ mid_x  # для каждого столбца: сумма n_ij * x_i

df_corr = pd.DataFrame(H, index=row_labels, columns=col_labels).astype(object)

# Заменяем нули на "—"
for i in range(k):
    for j in range(k):
        if H[i, j] == 0:
            df_corr.iloc[i, j] = '—'

# Добавляем столбцы с вычислениями
df_corr[r'$\sum n_{ij} \cdot y_j$'] = [f'{v:.3f}' for v in weighted_sum_y]
df_corr[r'$x_i \cdot \sum n_{ij} \cdot y_j$'] = [f'{mid_x[i] * weighted_sum_y[i]:.3f}' for i in range(k)]

# Добавляем строки с вычислениями для столбцов
row_y = [f'{v:.3f}' for v in weighted_sum_x] + ['', '']
row_xy = [f'{mid_y[j] * weighted_sum_x[j]:.3f}' for j in range(k)] + ['', '']

# Добавляем строки в DataFrame
df_corr.loc[r'$\sum n_{ij} \cdot x_i$'] = row_y
df_corr.loc[r'$y_j \cdot \sum n_{ij} \cdot x_i$'] = row_xy

# Вычисляем общую сумму для контрольной ячейки
total_sum = sum(mid_x[i] * weighted_sum_y[i] for i in range(k))
df_corr.iloc[k+1, k] = f'{total_sum:.3f}'  # правый нижний угол

df_corr.index.name = 'pH \\ sulphates'
display(Markdown(df_corr.to_markdown()))


print("1. Анализ распределения частот:")

# Находим максимальную частоту
max_freq_idx = np.unravel_index(np.argmax(H), H.shape)
max_freq = H[max_freq_idx]
print(f"   - Максимальная частота {max_freq} наблюдений приходится на")
print(f"     pH ≈ {mid_x[max_freq_idx[0]]:.3f} и sulphates ≈ {mid_y[max_freq_idx[1]]:.3f}")

print("\n2. Характер связи между признаками:")

# Вычисляем условные средние
cond_means_y = weighted_sum_y / row_sums
cond_means_x = weighted_sum_x / col_sums

# Проверяем монотонность
y_increasing = np.all(np.diff(cond_means_y) > 0)
y_decreasing = np.all(np.diff(cond_means_y) < 0)
x_increasing = np.all(np.diff(cond_means_x) > 0)
x_decreasing = np.all(np.diff(cond_means_x) < 0)

if y_increasing and x_increasing:
    print("   - Наблюдается прямая корреляционная связь")
elif y_decreasing and x_decreasing:
    print("   - Наблюдается обратная корреляционная связь")
else:
    print("   - Явно выраженная корреляционная связь не прослеживается")

print("\n3. Особенности распределения:")
print("   - По строкам (фиксированный pH) видно распределение sulphates")
print("   - По столбцам (фиксированный sulphates) видно распределение pH")
print("   - Наличие нулевых частот указывает на отсутствие наблюдений в соответствующих комбинациях интервалов")

| pH \ sulphates                    | 0.514   | 0.661   | 0.809   | 0.956   | 1.104   | 1.251   | 1.399   | 1.546   | $\sum n_{ij} \cdot y_j$   | $x_i \cdot \sum n_{ij} \cdot y_j$   |
|:----------------------------------|:--------|:--------|:--------|:--------|:--------|:--------|:--------|:--------|:--------------------------|:------------------------------------|
| 2.925                             | 2       | —       | 1       | —       | —       | —       | 1       | 1       | 4.781                     | 13.985                              |
| 3.055                             | 4       | 1       | —       | 1       | 1       | —       | 1       | —       | 6.175                     | 18.865                              |
| 3.185                             | 8       | 11      | 6       | 1       | —       | —       | —       | —       | 17.193                    | 54.758                              |
| 3.315                             | 18      | 11      | 7       | —       | —       | 2       | —       | —       | 24.685                    | 81.831                              |
| 3.445                             | 10      | 14      | 2       | —       | —       | —       | —       | —       | 16.013                    | 55.163                              |
| 3.575                             | 2       | 4       | —       | —       | —       | —       | —       | —       | 3.673                     | 13.129                              |
| 3.705                             | 1       | —       | 2       | —       | —       | —       | —       | —       | 2.131                     | 7.896                               |
| 3.835                             | —       | 2       | —       | —       | —       | —       | —       | —       | 1.323                     | 5.072                               |
| $\sum n_{ij} \cdot x_i$           | 148.525 | 144.755 | 59.540  | 6.240   | 3.055   | 6.630   | 5.980   | 2.925   |                           |                                     |
| $y_j \cdot \sum n_{ij} \cdot x_i$ | 76.305  | 95.719  | 48.153  | 5.967   | 3.372   | 8.296   | 8.365   | 4.523   | 250.699                   |                                     |

1. Анализ распределения частот:
   - Максимальная частота 18 наблюдений приходится на
     pH ≈ 3.315 и sulphates ≈ 0.514

2. Характер связи между признаками:
   - Явно выраженная корреляционная связь не прослеживается

3. Особенности распределения:
   - По строкам (фиксированный pH) видно распределение sulphates
   - По столбцам (фиксированный sulphates) видно распределение pH
   - Наличие нулевых частот указывает на отсутствие наблюдений в соответствующих комбинациях интервалов


### **4.** Исходя из результатов корреляционной таблицы вычислить значение выборочного коэффициента корреляции двумя способами: с помощью стандартной формулы и с помощью условных вариант. Убедиться, что результаты совпадают. Сделать выводы.

In [ ]:
# Способ 1 - стандартная формула по исходным данным
r_standard = np.corrcoef(X, Y)[0, 1]

# Способ 2 - через условные варианты
h_x = mid_x[1] - mid_x[0]
h_y = mid_y[1] - mid_y[0]
C_x = mid_x[k // 2]
C_y = mid_y[k // 2]

# Условные варианты u и v
u = (mid_x - C_x) / h_x
v = (mid_y - C_y) / h_y

row_sums = H.sum(axis=1)
col_sums = H.sum(axis=0)
u_bar = np.dot(row_sums, u) / n
v_bar = np.dot(col_sums, v) / n

sum_nuv = sum(
    H[i, j] * u[i] * v[j]
    for i in range(k) for j in range(k)
)

# σ̄_u² = Σ nᵢ. · uᵢ² / N − ū_в²
sigma_u = np.sqrt(np.dot(row_sums, u**2) / n - u_bar**2)
# σ̄_v² = Σ n.ⱼ · vⱼ² / N − v̄_в²
sigma_v = np.sqrt(np.dot(col_sums, v**2) / n - v_bar**2)

# r̄_xy по формуле учебника
r_table = (sum_nuv - n * u_bar * v_bar) / (n * sigma_u * sigma_v)

print(f"Способ 1 (стандартная формула):     r = {r_standard:.6f}")
print(f"Способ 2 (через условные варианты): r = {r_table:.6f}")
print(f"Расхождение: Δr = {abs(r_standard - r_table):.6f}")

Способ 1 (стандартная формула):     r = -0.281390
Способ 2 (через условные варианты): r = -0.248692
Расхождение: Δr = 0.032698


Выборочный коэффициент корреляции вычислен двумя способами. По стандартной формуле получено $r = -0.281390$, по методу условных вариант $r = -0.248692$. Расхождение между результатами составляет $\Delta r = 0.032698$, что объясняется потерей информации при группировке данных в интервальный ряд: при переходе от исходных индивидуальных значений к серединам интервалов происходит усреднение данных внутри каждого интервала, что приводит к некоторому сглаживанию вариации и, как следствие, к незначительному смещению оценки коэффициента корреляции. Тем не менее, оба метода дают согласованные результаты, указывающие на одну и ту же направленность и силу связи.

Величина $|r| = 0.281390$ соответствует слабой линейной корреляционной зависимости. Знак «минус» указывает на отрицательную направленность связи, то есть с увеличением $pH$ незначительно снижается $sulphates$, и наоборот.

### **5.** Построить доверительный интервал для коэффициента корреляции при уровне значимости $\gamma \in {0.95,0.99}$, сделать выводы.

In [ ]:
from scipy import stats
# Шаг 1: z̄ по формуле
z_bar = 0.5 * np.log((1 + r_standard) / (1 - r_standard))
print(f"z̄ = {z_bar:.4f}")

# Шаг 2: σ̄_z по формуле
sigma_z = 1 / np.sqrt(n - 3)
print(f"σ̄_z = {sigma_z:.4f}")

for gamma in [0.95, 0.99]:
    # Шаг 3: λ(γ) из условия Φ[λ(γ)] = γ/2
    lam = stats.norm.ppf(0.5 + gamma / 2)

    # ДИ в z-шкале по формуле
    z_low  = z_bar - lam * sigma_z
    z_high = z_bar + lam * sigma_z

    # Шаг 4: обратное преобразование Фишера
    r_low  = (np.exp(2 * z_low)  - 1) / (np.exp(2 * z_low)  + 1)
    r_high = (np.exp(2 * z_high) - 1) / (np.exp(2 * z_high) + 1)

    print(f"\nγ = {gamma}:")
    print(f"\tλ(γ) = {lam:.4f}  [из Φ(λ) = {gamma}/2 = {gamma/2}]")
    print(f"\tДИ в z-шкале: ({z_low:.4f}; {z_high:.4f})")
    print(f"\tДИ для r_xy: ({r_low:.4f};  {r_high:.4f})")

z̄ = -0.2892
σ̄_z = 0.0949

γ = 0.95:
	λ(γ) = 1.9600  [из Φ(λ) = 0.95/2 = 0.475]
	ДИ в z-шкале: (-0.4752; -0.1032)
	ДИ для r_xy: (-0.4424;  -0.1028)

γ = 0.99:
	λ(γ) = 2.5758  [из Φ(λ) = 0.99/2 = 0.495]
	ДИ в z-шкале: (-0.5337; -0.0447)
	ДИ для r_xy: (-0.4882;  -0.0447)



Доверительный интервал для генерального коэффициента корреляции построен методом
z-преобразования Фишера. По выборочному значению $r = -0.2814$ вычислено
$\bar{z} = -0.2892$ со стандартной ошибкой $\bar{\sigma_z} = 0.0949$.

При доверительной вероятности $\gamma = 0.95$ доверительный интервал для $r_{xy}$ составил ($-0.4424$; $-0.1028$). Это означает, что с вероятностью 95% истинное значение коэффициента корреляции между $pH$ и $sulphates$ в генеральной совокупности находится в указанных пределах. Поскольку весь интервал не содержит ноля и расположен в области отрицательных значений, можно утверждать о наличии статистически значимой обратной связи между исследуемыми признаками.

При доверительной вероятности $\gamma = 0.99$ доверительный интервал расширяется до ($-0.4882$; $-0.0447$), что соответствует большей надежности оценки, но меньшей точности. В данном интервале также отсутствует ноль, что подтверждает значимость корреляционной связи даже при более строгом уровне значимости.

Таким образом, с увеличением доверительной вероятности ширина интервала
возрастает (от $0.3396$ при $\gamma = 0.95$ до $0.4435$ при $\gamma = 0.99$), что
соответствует теоретическим положениям: чем выше требуемая надежность оценки,
тем шире доверительный интервал.

Отрицательный знак коэффициента корреляции свидетельствует об обратной
зависимости: с увеличением $pH$ содержание сульфатов
имеет тенденцию к уменьшению, либо наоборот.

### **6.** Осуществить проверку статистической гипотезы о равенстве коэффициента корреляции нулю при заданном уровне значимости $\alpha = 0.05$, сделать выводы.

In [ ]:
alpha = 0.05
t_obs  = r_standard * np.sqrt(n - 2) / np.sqrt(1 - r_standard**2)
t_crit = stats.t.ppf(1 - alpha / 2, df=n - 2)

print(f"r = {r_standard:.4f}")
print(f"n = {n}")
print(f"T_набл = {t_obs:.4f}")
print(f"T_крит = ±{t_crit:.4f}  (df={n-2}, α={alpha})")
print(f"|T_набл| = {abs(t_obs):.4f} {'>' if abs(t_obs) > t_crit else '<'} {t_crit:.4f} = T_крит")

print(f"p-value = {2 * (1 - stats.t.cdf(abs(t_obs), df=n-2)):.4f}")

r = -0.2814
n = 114
T_набл = -3.1033
T_крит = ±1.9814  (df=112, α=0.05)
|T_набл| = 3.1033 > 1.9814 = T_крит
p-value = 0.0024


При проверке статистической гипотезы $H_0: r_{xy} = 0$ при уровне значимости $\alpha = 0.05$ наблюдаемое значение статистики составило $|T_{набл}| = 3.1033$, что превышает критическое значение $T_{крит} = 1.9814$. Следовательно, нулевая гипотеза $H_0$ отвергается.

Это подтверждается и значением $p-value = 0.0024$: оно меньше уровня значимости $\alpha = 0.05$.

Таким образом, на уровне значимости $0.05$ установлено, что между признаками $pH$ и $suplhates$ существует статистически значимая отрицательная линейная корреляция $(r = −0.2814$).

## **Вывод**


В ходе выполнения практической работы освоены основные понятия, связанные с корреляционной зависимостью между случайными величинами, а также методы проверки статистических гипотез об их «справедливости».

На первом этапе на основе двумерного интервального вариационного ряда построена корреляционная таблица, позволившая визуализировать совместное распределение признаков $pH$ и $sulphates$ и выявить предварительный характер связи между ними.

Далее двумя способами - по стандартной формуле и методом условных вариант - вычислен выборочный коэффициент корреляции Пирсона. Получены значения $( r = -0.2814 )$ и $( r = -0.2487)$ соответственно. Расхождение между результатами $( \Delta r = 0.032698)$ обусловлено неизбежной потерей информации при группировке исходных данных в интервальный ряд, что подтверждает важность работы с исходными данными при точных расчетах. Оба метода дали согласованные результаты, указывающие на слабую отрицательную связь между исследуемыми признаками.

Для оценки надежности полученного коэффициента корреляции построены доверительные интервалы с использованием Z-преобразования Фишера при доверительных вероятностях $( \gamma = 0.95)$ и $( \gamma = 0.99)$. При $( \gamma = 0.95)$ доверительный интервал составил ($-0.4424$; $-0.1028$), что не содержит нуля и свидетельствует о статистической значимости связи на данном уровне. При $( \gamma = 0.99)$ интервал ($-0.4882$; $-0.0447$) также не  содержит ноль.

Заключительным этапом работы стала проверка статистической гипотезы о равенстве коэффициента корреляции нулю при уровне значимости $( \alpha = 0.05)$. Рассчитанное наблюдаемое значение $( T_{\text{набл}} = -3.1033)$ превысило критическое значение $( T_{\text{крит}} = \pm1.9814)$, а $( p\text{-value} = 0.0024)$ оказалось меньше уровня значимости. Это позволило отвергнуть нулевую гипотезу $( H_0: r = 0)$ и сделать вывод о статистической значимости обнаруженной корреляционной связи.

Таким образом, в результате выполнения работы практически освоены: построение корреляционной таблицы, вычисление коэффициента корреляции различными методами, построение доверительных интервалов с использованием Z-преобразования Фишера, проверка статистических гипотез о значимости коэффициента корреляции с применением t-критерия и анализа p-value. Полученные навыки позволяют корректно интерпретировать наличие, направление и силу корреляционной связи между случайными величинами, а также принимать обоснованные статистические решения о значимости выявленных зависимостей.